packages/gatsby-source-contentful/src/gatsby-node.js

```diff
@@ -0,0 +1,174 @@
+const contentful = require(`contentful`)
+const crypto = require(`crypto`)
+const _ = require(`lodash`)
+
+const typePrefix = `contentful__`
+const makeTypeName = type => `${typePrefix}${type.replace(/-/g, `_`)}`
+
+exports.sourceNodes = async (
+  { boundActionCreators, getNode, hasNodeChanged, store },
+  { spaceId, accessToken }
+) => {
+  const {
+    createNode,
+    updateSourcePluginStatus,
+    touchNode,
+  } = boundActionCreators
+  updateSourcePluginStatus({
+    plugin: `gatsby-source-contentful`,
+    status: {
+      ...store.getState().status[`gatsby-source-contentful`],
+      ready: false,
+    },
+  })
+
+  // Touch existing Contentful nodes so Gatsby doesn't garbage collect them.
+  _.values(store.getState().nodes)
+    .filter(n => n.internal.type.slice(0, 8) === typePrefix)
+    .forEach(n => touchNode(n.id))
+
+  // Fetch articles.
+  console.time(`fetch Contentful data`)
+  console.log(`Starting to fetch data from Contentful`)
+
+  const client = contentful.createClient({
+    space: spaceId,
+    accessToken,
+  })
+
+  let contentTypes
+  try {
+    contentTypes = await client.getContentTypes()
+  } catch (e) {
+    console.log(`error fetching content types`, e)
+  }
+  console.log(`contentTypes fetched`, contentTypes.items.length)
+
+  const entryList = await Promise.all(contentTypes.items.map(async (contentType) => {
+    const contentTypeId = contentType.sys.id
+    let entries
+    try {
+      entries = await client.getEntries({
+        content_type: contentTypeId,
+      })
+    } catch (e) {
+      console.log(`error fetching entries`, e)
+    }
+    console.log(`entries fetched for content type ${contentType.name} (${contentTypeId})`, entries.items.length)
+    return entries
+  }))
+
+  let assets
+  try {
+    assets = await client.getAssets()
+  } catch (e) {
+    console.log(`error fetching assets`, e)
+  }
+  console.log(`assets fetched`, assets.items.length)
+
+  updateSourcePluginStatus({
+    plugin: `gatsby-source-contentful`,
+    status: {
+      ...store.getState().status[`gatsby-source-contentful`],
+      lastFetched: new Date().toJSON(),
+    },
+  })
+
+  console.timeEnd(`fetch Contentful data`)
+
+  const contentTypeItems = contentTypes.items
+  contentTypeItems.forEach((contentTypeItem, i) => {
+    const contentTypeItemId = contentTypeItem.sys.id
+
+    // First create nodes for each of the entries of that content type
+    const entryNodes = entryList[i].items.map((entryItem) => {
+
+      const entryNode = {
+        id: entryItem.sys.id,
+        parent: contentTypeItemId,
+        children: [],
+        name: entryItem.name,
+        ...entryItem.fields,
```

Comment:
I think our only choice is to prefix any conflicting fields. And tell people when designing their schemas to avoid conflicting if possible with Gatsby's fields.